security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -0,0 +1,241 @@
+}
+
+type sdsConnection struct {
+	// PeerAddr is the address of the client envoy, from network layer.
+	PeerAddr string
+
+	// Time of connection, for debugging.
+	Connect time.Time
+
+	// The proxy from which the connection comes from.
+	modelNode *model.Proxy
+
+	// doneChannel will be closed when the client is closed.
+	doneChannel chan int
+
+	// SDS streams implement this interface.
+	stream discoveryStream
+}
+
+type sdsservice struct {
+	st secretStore
+	//TODO(quanlin), add below properties later:
+	//1. workloadRegistry(store proxies information).
+	//2. caClient(interact with CA for CSR).
+}
+
+// newSDSService creates Secret Discovery Service which implements envoy v2 SDS API.
+func newSDSService(st secretStore) *sdsservice {
+	return &sdsservice{
+		st: st,
+	}
+}
+
+// register adds the SDS handle to the grpc server
+func (s *sdsservice) register(rpcs *grpc.Server) {
+	sds.RegisterSecretDiscoveryServiceServer(rpcs, s)
+}
+
+func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecretsServer) error {
+	peerAddr := "Unknown peer address"
+	peerInfo, ok := peer.FromContext(stream.Context())
+	if ok {
+		peerAddr = peerInfo.Addr.String()
+	}
+
+	var discReq *xdsapi.DiscoveryRequest
+	var receiveError error
+	reqChannel := make(chan *xdsapi.DiscoveryRequest, 1)
+
+	con := newSDSConnection(peerAddr, stream)
+	defer close(con.doneChannel)
+
+	go receiveThread(con, reqChannel, &receiveError)
+
+	for {
+		// Block until either a request is received.
+		select {
+		case discReq, ok = <-reqChannel:
+			if !ok {
+				// Remote side closed connection.
+				return receiveError
+			}
+
+			if discReq.Node.Id == "" {
+				log.Warnf("Discovery request %+v missing node id", discReq)
+				continue
+			}
+			nt, err := model.ParseServiceNode(discReq.Node.Id)
+			if err != nil {
+				log.Errorf("Failed to parse service node from discovery request %+v: %v", discReq, err)
+				return err
+			}
+			nt.Metadata = model.ParseMetadata(discReq.Node.Metadata)
+			con.modelNode = &nt
+
+			secret := s.st.getSecret()
+			if err := s.pushSds(secret, *con.modelNode, con); err != nil {
+				log.Errorf("SDS failed to push: %v", err)
+				return err
+			}
+		}
+	}
+}
+
+func (s *sdsservice) FetchSecrets(ctx context.Context, discReq *xdsapi.DiscoveryRequest) (*xdsapi.DiscoveryResponse, error) {
+	if discReq.Node.Id == "" {
+		log.Warnf("SDS discovery request %+v missing node id", discReq)
+		return nil, fmt.Errorf("SDS discovery request %+v missing node id", discReq)
+	}
+
+	proxy, err := model.ParseServiceNode(discReq.Node.Id)
+	if err != nil {
+		log.Errorf("Failed to parse service node from discovery request %+v: %v", discReq, err)
+		return nil, err
+	}
+
+	proxy.Metadata = model.ParseMetadata(discReq.Node.Metadata)
+
+	//TODO(quanlin): add proxy info in workload registry.
+
+	return sdsDiscoveryResponse(s.st.getSecret(), proxy)
+}
+
+func (s *sdsservice) pushSds(secret *secret, proxy model.Proxy, con *sdsConnection) error {
```

Comment:
Go encourage to use all capital case for acronym terms.  Let's be consistent by using "SDS", not "Sds".
https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Reply:
done.